File: frontend/src/queries/obsidian.ts
/** Obsidian queries + mutations. */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { api } from '@/lib/api';

export interface ObsidianStatus {
  enabled: boolean;
  vault_path: string;
  sync_mode: 'export_only' | 'two_way' | 'manual_only';
  template_set: 'complete' | 'minimal' | 'tactical';
  last_sync_at: string | null;
  last_error: string | null;
  conflict_count: number;
}

export interface ObsidianConflict {
  id: string;
  path: string;
  trade_id: string | null;
  detected_at: string;
  db_updated_at: string | null;
  file_updated_at: string | null;
  db_text: string | null;
  file_text: string | null;
}

export interface ExportSummary {
  trades_exported: number;
  daily_exported: number;
  strategies_exported: number;
  dashboards_exported: number;
  took_ms: number;
}

export interface ChangeEvent {
  path: string;
  trade_id: string | null;
  action: 'updated' | 'conflict' | 'unknown';
  detail: string | null;
}

export function useObsidianStatus(opts?: { refetchInterval?: number }) {
  return useQuery({
    queryKey: ['obsidian', 'status'],
    queryFn: () => api.get<ObsidianStatus>('/obsidian/status'),
    refetchInterval: opts?.refetchInterval,
  });
}

export function usePatchObsidianConfig() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: Partial<Pick<ObsidianStatus, 'enabled' | 'vault_path' | 'sync_mode' | 'template_set'>>) =>
      api.patch<ObsidianStatus>('/obsidian/config', body),
    onSuccess: (data) => {
      qc.setQueryData(['obsidian', 'status'], data);
    },
  });
}

export function useExportAll() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: () => api.post<ExportSummary>('/obsidian/export-all'),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['obsidian'] }),
  });
}

export function useSyncNow() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: () => api.post<ChangeEvent[]>('/obsidian/sync-now'),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['obsidian'] });
      qc.invalidateQueries({ queryKey: ['trades'] });
    },
  });
}

export function useConflicts() {
  return useQuery({
    queryKey: ['obsidian', 'conflicts'],
    queryFn: () => api.get<ObsidianConflict[]>('/obsidian/conflicts'),
  });
}

export function useResolveConflict() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (args: { id: string; resolution: 'keep_db' | 'keep_file' | 'merged'; merged_text?: string }) =>
      api.post(`/obsidian/conflicts/${args.id}/resolve`, {
        resolution: args.resolution,
        merged_text: args.merged_text,
      }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['obsidian'] }),
  });
}

File: backend/app/schemas/obsidian.py
"""Obsidian config + status schemas."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


SyncMode = Literal["export_only", "two_way", "manual_only"]
TemplateSet = Literal["complete", "minimal", "tactical"]


class ObsidianStatus(BaseModel):
    enabled: bool
    vault_path: str
    sync_mode: SyncMode
    template_set: TemplateSet
    last_sync_at: datetime | None
    last_error: str | None
    conflict_count: int


class ObsidianConfigUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    enabled: bool | None = None
    vault_path: str | None = Field(default=None, max_length=512)
    sync_mode: SyncMode | None = None
    template_set: TemplateSet | None = None


class ExportSummary(BaseModel):
    trades_exported: int
    daily_exported: int
    strategies_exported: int
    dashboards_exported: int
    took_ms: int


class ChangeEvent(BaseModel):
    path: str
    trade_id: str | None
    action: Literal["updated", "conflict", "unknown"]
    detail: str | None = None


class ConflictOut(BaseModel):
    id: uuid.UUID
    path: str
    trade_id: uuid.UUID | None
    detected_at: datetime
    db_updated_at: datetime | None
    file_updated_at: datetime | None
    db_text: str | None
    file_text: str | None

    model_config = ConfigDict(from_attributes=True)


class ConflictResolveRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    resolution: Literal["keep_db", "keep_file", "merged"]
    merged_text: str | None = None
